packages/mcp-auth/src/auth/resource-server-handler.ts

```diff
@@ -0,0 +1,100 @@
+import { type Router } from 'express';
+
+import { MCPAuthAuthServerError } from '../errors.js';
+import { createResourceMetadataRouter } from '../routers/create-resource-metadata-router.js';
+import { type ResourceServerConfig } from '../types/resource-server.js';
+import { transpileResourceMetadata } from '../utils/transpile-resource-metadata.js';
+import { validateAuthServer } from '../utils/validate-auth-server.js';
+
+import { MCPAuthHandler } from './mcp-auth-handler.js';
+import { type GetTokenVerifierOptions, TokenVerifier } from './token-verifier.js';
+
+/**
+ * Configuration for the MCP server as resource server mode.
+ */
+export type ResourceServerModeConfig = {
+  /**
+   * A single resource server configuration or an array of them.
+   */
+  protectedResources: ResourceServerConfig | ResourceServerConfig[];
+};
+/**
+ * Handles the authentication logic for the MCP server as resource server mode.
+ */
+export class ResourceServerHandler extends MCPAuthHandler {
+  /** A map of `TokenVerifier` instances for each configured resource server, keyed by resource identifier. */
+  private readonly tokenVerifiers: Map<string, TokenVerifier>;
+
+  constructor(private readonly config: ResourceServerModeConfig) {
+    super();
+    this.#validateConfig(this.resourcesConfigs);
+
+    this.tokenVerifiers = new Map();
+    for (const resourceConfig of this.resourcesConfigs) {
+      const {
+        metadata: { resource, authorizationServers },
+      } = resourceConfig;
+      this.tokenVerifiers.set(resource, new TokenVerifier(authorizationServers ?? []));
+    }
+  }
+
+  createMetadataRouter(): Router {
+    return createResourceMetadataRouter(
+      this.resourcesConfigs.map((config) => transpileResourceMetadata(config.metadata))
+    );
+  }
+
+  getTokenVerifier(options?: GetTokenVerifierOptions): TokenVerifier {
+    const { resource } = options ?? {};
+
+    if (!resource) {
+      throw new MCPAuthAuthServerError('invalid_server_config', {
+        cause:
+          'A `resource` must be specified in the `bearerAuth` configuration when using a `protectedResources` configuration.',
+      });
+    }
```

Comment:
if protected resource is the new norm, will it be better to make `options` and `options.resource` required in the abstract class, and ignore or use a placeholder in `AuthorizationServerHandler`?